### DISCUSSION TITLE
Create table as select (CTAS)

### DISCUSSION THREAD

**Discussion by fabiomx** [littleK0i/SnowDDL#49] [General]
Hello! 

I would like to create a table using the "CREATE TABLE AS SELECT" (CTAS) statement, as mentioned in the [Snowflake documentation](https://docs.snowflake.com/en/sql-reference/sql/create-table), but I can't see the option in the Table YAML config page ([TABLE](https://docs.snowddl.com/basic/yaml-configs/table)). 

On the other hand, this option seems to be mentioned at the end of the page:

> Additionally, in order to execute CREATE OR REPLACE TABLE ... AS SELECT , an additional CLI option must be specified:
 --apply-replace-table

What am I missing? 

I would expect a text field similar to the one available for materialized views or dynamic tables. 

Thanks!

**Comment by littleK0i** [comment]
`CREATE OR REPLACE TABLE` is produced by SnowDDL when table structure was changed, but it cannot be rectified by `ALTER TABLE ... ALTER COLUMN`. Snowflake requires full table rewrite for most transformations, and this is when you may need option `--apply-replace-table`.

More details: https://docs.snowflake.com/en/sql-reference/sql/alter-table-column

Normal `TABLE` object cannot be created from SELECT by design. Reasons:

1) You probably want not only to create table once, but to refresh it according to schedule or some external trigger.
2) There is no way to compare table created by CTAS with table currently existing in DWH without actually running the CTAS every time, which might be very expensive.

What you can use instead:

- Define YAML for such table normally (with `columns`). Schedule an ETL job (e.g. Airflow, DBT) refreshing this table. You may use `INSERT OVERWRITE INTO <table> AS SELECT ...` command.
- Use `MATERIALIZED_VIEW`.
- Use `DYNAMIC_TABLE` with schedule.
- Use `TASK` with schedule running `PROCEDURE` which updates the table.

The first option is preferred, since it provided the most flexibility.

**Comment by fabiomx** [comment]
I would like to provide additional context to help you better understand why we are considering using a CTAS statement and why the options you mentioned would not be suitable, in my opinion, as they would involve unnecessary complexity.

Firstly, I want to clarify that we do not require or desire to refresh this particular table. The select operation is performed on legacy data that is no longer being updated.

Using a MATERIALIZED VIEW is not an option for us because we need to utilize this table alongside other tables that are regularly refreshed, and they will be used to generate a dynamic table. Even if the dynamic table is only relevant for the new tables and not the legacy one, all the tables are necessary for creating the final aggregated data table. Unfortunately, since dynamic tables cannot be built based on materialized views, it is not feasible to create the legacy table as a materialized view.

As far as I know, one potential solution could involve retrieving the definition of the TABLE (as CTAS) and performing a comparison with the one available in the YAML. Is it feasible to implement this approach?

Thank you!

**Comment by littleK0i** [comment]
Aha, I see.

This specific problem is slightly out of scope for SnowDDL, since it is an "object management tool". It manages table definitions and object schema. It does not manage table contents and data.

But you can achieve the desired goal using the following simple approach:

- Create a YAML config for table with legacy data.
- Run `snowddl apply`. An empty table will be created.
- Run `INSERT INTO <legacy_data_table> SELECT ...` once manually.

Legacy data table will be filled with snapshot of data. It is almost the same thing you would get with `CREATE TABLE ... AS SELECT`.

Also, it might be a good idea to create a backup of this data somewhere in S3 / Cloud Storage / Blob Storage using [COPY INTO ... LOCATION](https://docs.snowflake.com/en/sql-reference/sql/copy-into-location) command, just in case.

Otherwise you may lose this data at some point in future, if someone drops or modifies this table accidentally.

**Comment by littleK0i** [comment]
About other object types with `text` parameter.

- `VIEW` is just a definition, no real data is being moved around;
- data in `DYNAMIC_TABLE` is managed by Snowflake in background;
- data in `MATERIALIZED_VIEW` is managed by Snowflake in background;
- `TASKS` are executed by Snowflake in background;

All DDL operations on these object types are instant.

But `CREATE OR REPLACE TABLE ... AS SELECT` is a DML operation which actually writes data. Depending on the size of data and complexity of query, it may take several hours to finish.

**Comment by fabiomx** [comment]
I see. Thanks for your answers.

Just let me add that even if a CTAS incorporates DML operations, it is usually classified as DDL (by Snowflake too) since it creates a table based on the query result.

Because of that, I would like to keep a record of that query somewhere, and I think that the SnowDDL folder would be the most appropriate place. Following your suggested approach, if I leave a SQL file containing the query used to run the insert in the SnowDDL configuration folder (in the `table` folder), will it be safely ignored by SnowDDL?

Thanks!

**Reply by littleK0i** [reply]
All non-YAML files will be ignored by config parsers, so it safe.

However, I would generally suggest to create a separate directory somewhere completely outside of SnowDDL config and create 1 file per major release with manual SQL commands applied during this release.

You may have releases with a large number of various transformations in future. It's much easier when such SQL is logically grouped together and explicitly attached to specific version.

**Reply by fabiomx** [reply]
We are still exploring options, however, this solution would be just for this specific and unusual case, a one-time DDL/DML script. 

Standard pipelines and transformations would be handled and versioned outside of SnowDDL.

Thanks!
